runtime/onert/backend/train/ops/FullyConnectedLayer.cc

```diff
@@ -0,0 +1,179 @@
+    _transposed_input{nullptr}, _transposed_deriv_output{nullptr}, _act_deriv_output{nullptr}
+{
+  // DO NOTHING
+}
+
+FullyConnectedLayer::~FullyConnectedLayer() = default;
+
+void FullyConnectedLayer::configure(const IPortableTensor *input, const IPortableTensor *weights,
+                                    const IPortableTensor *bias, IPortableTensor *output,
+                                    IPortableTensor *deriv_input, IPortableTensor *grad_weights,
+                                    IPortableTensor *grad_bias, const IPortableTensor *deriv_output,
+                                    ir::Activation activation,
+                                    ir::FullyConnectedWeightsFormat weights_format,
+                                    const std::shared_ptr<train::ExternalContext> &external_context)
+{
+  cpu::ops::FullyConnectedLayer::configure(input, weights, bias, activation, weights_format, output,
+                                           external_context);
+
+  _deriv_input = deriv_input;
+  _grad_weights = grad_weights;
+  _grad_bias = grad_bias;
+  _deriv_output = deriv_output;
+
+  if (weights_format != ir::FullyConnectedWeightsFormat::Default)
+    throw std::runtime_error{
+      "train FullyConnected: Weight formats other than default are not supported."};
+
+  if (input->get_info().shape().rank() != 2 || weights->get_info().shape().rank() != 2 ||
+      output->get_info().shape().rank() != 2 || deriv_input->get_info().shape().rank() != 2 ||
+      grad_weights->get_info().shape().rank() != 2 || deriv_output->get_info().shape().rank() != 2)
+    throw std::runtime_error{"train FullyConnected: Input other ranks than 2 are not supported."};
+
+  _transposed_weights = std::make_unique<Tensor>(weights->get_info(), weights->layout());
+  _transposed_weights->setBuffer(std::make_shared<basic::Allocator>(weights->total_size()));
+
+  _transposed_input = std::make_unique<Tensor>(input->get_info(), input->layout());
+  _transposed_input->setBuffer(std::make_shared<basic::Allocator>(input->total_size()));
+
+  _transposed_deriv_output =
+    std::make_unique<Tensor>(deriv_output->get_info(), deriv_output->layout());
+  _transposed_deriv_output->setBuffer(
+    std::make_shared<basic::Allocator>(deriv_output->total_size()));
+
+  if (activation != ir::Activation::NONE)
+  {
+    _act_deriv_output =
+      std::make_unique<Tensor>(_deriv_output->get_info(), _deriv_output->layout());
+    _act_deriv_output->setBuffer(std::make_shared<basic::Allocator>(_deriv_output->total_size()));
+  }
+}
+
+void FullyConnectedLayer::forward(bool) { cpu::ops::FullyConnectedLayer::run(); }
+
+void FullyConnectedLayer::backward(uint32_t)
+{
+  const auto data_type = _deriv_output->data_type();
+  assert(data_type == _input->data_type());
+  switch (data_type)
+  {
+    case OperandType::FLOAT32:
+    {
+      assert(data_type == _grad_weights->data_type());
+      assert(data_type == _grad_bias->data_type());
```

Comment:
Can we move these asserts in front of the switch phrase?

Reply:
There are cases the data type of weights and bias is different from the data type of input, so it looks better to put them here.